nexus/codec.go

```diff
@@ -0,0 +1,204 @@
+	// User provided keys are considered case-insensitive by the framework.
+	Header map[string]string
+	// Reader contains request or response data. May be nil for empty data.
+	Reader io.ReadCloser
+}
+
+// A LazyValue holds a value encoded in an underlying [Stream].
+//
+// ⚠️ When a LazyValue is returned from a client - if directly accessing the stream - it must be read it in its entirety
+// and closed to free up the associated HTTP connection. Otherwise the [LazyValue.Consume] method must be called.
+//
+// ⚠️ When a LazyValue is passed to a server handler, it must not be used after the returning from the handler method.
+type LazyValue struct {
+	codec  Codec
+	Stream *Stream
+}
+
+// Consume consumes the lazy value, decodes it from the underlying stream, and stores the result in the value pointed to by v.
+//
+//	var v int
+//	err := s.Consume(&v)
+func (s *LazyValue) Consume(v any) error {
+	defer s.Stream.Reader.Close()
+	return s.codec.Decode(s.Stream, v)
+}
+
+// Codec is used by the framework to serialize/deserialize input and output.
+// To customize serialization logic, implement this interface and provide your implementation to framework methods such
+// as [NewClient] and [NewHTTPHandler].
+// By default, the SDK supports serialization of JSONables, byte slices, and nils.
+type Codec interface {
+	// Encode encodes a value into a [Stream].
+	Encode(any) (*Stream, error)
+	// Decode decodes a [Stream] into a given reference.
+	Decode(*Stream, any) error
+}
+
+type codecChain []Codec
+
+var errCodecIncompatible = errors.New("incompatible codec")
+
+func (c codecChain) Encode(v any) (*Stream, error) {
+	for _, l := range c {
+		p, err := l.Encode(v)
+		if err != nil {
+			if errors.Is(err, errCodecIncompatible) {
+				continue
+			}
+			return nil, err
+		}
+		return p, nil
+	}
+	return nil, errCodecIncompatible
+}
+
+func (c codecChain) Decode(s *Stream, v any) error {
+	lenc := len(c)
+	for i := range c {
+		l := c[lenc-i-1]
+		if err := l.Decode(s, v); err != nil {
+			if errors.Is(err, errCodecIncompatible) {
+				continue
+			}
+			return err
+		}
+		return nil
+	}
+	return errCodecIncompatible
+}
+
+var _ Codec = codecChain{}
+
+type jsonCodec struct{}
+
+func (jsonCodec) Decode(s *Stream, v any) error {
+	if !isMediaTypeJSON(s.Header["content-type"]) {
+		return errCodecIncompatible
+	}
+	body, err := io.ReadAll(s.Reader)
+	if err != nil {
+		return err
+	}
+	return json.Unmarshal(body, &v)
+}
+
+func (jsonCodec) Encode(v any) (*Stream, error) {
+	body, err := json.Marshal(v)
+	if err != nil {
+		return nil, err
+	}
+	return &Stream{
+		Header: map[string]string{
+			"content-type":   "application/json",
+			"content-length": fmt.Sprintf("%d", len(body)),
+		},
+		Reader: io.NopCloser(bytes.NewReader(body)),
+	}, nil
+}
+
+var _ Codec = jsonCodec{}
+
+type nilCodec struct{}
+
+func (nilCodec) Decode(s *Stream, v any) error {
+	if s.Header["content-length"] != "0" {
```

Comment:
Be advised, not all HTTP clients set a content length for their requests. What happens if I make an empty Nexus call with no content type, no content length, etc? Does Nexus reject such a simple request?

Reply:
TBH, I'm not sure what to do here.

I guess I could try and read a byte of content and see if I get EOF and if it's not empty, mutate the reader to concatenate the bytes read with the rest of the content since the reader might not be seekable.
It just seems to complex and not worth it.

I also don't think I want to rely on content type here either.

Reply:
We may be able to step back and ask why nil is needed/important as it's the absence of content. If for the JSON and byte slice converters we need to support absence of content we can. Can you show a situation where this is needed?

The way I am starting to think about it, only the people that actually read content should care about nil or not, and only once reading the content will that matter.

But yes, technically you could make a `bufio.NewReader` and invoke `Peek`.

Reply:
I don't think we should mutate the content in the serializers (note using new terminology).

For now I think we should force setting a specific content type or content length 0.

Reply:
I think we should support simple content-less HTTP calls. Maybe we shouldn't support optional JSON input. `[]byte` can just read 0 bytes and be fine, this whole thing is only about nullable JSON objects at this point and it's not right to make nullable JSON object be the reason that everyone must set some header even if they use `[]byte` or no parameter at all.